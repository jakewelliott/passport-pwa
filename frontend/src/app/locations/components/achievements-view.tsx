import type { Park, ParkActivity } from '@/lib/mock/types';
import { FaRegCheckSquare, FaRegSquare, FaStamp } from 'react-icons/fa';

interface AchievementsViewProps {
  parkActivity: ParkActivity;
  park: Park;
}

const AchievementsView = ({ parkActivity, park }: AchievementsViewProps) => {
  return (
    <div className='flex flex-col gap-3' data-testid='achievements-view'>
      {parkActivity.stampCollectedAt !== undefined && (
        <div className='top-0 flex'>
          <FaStamp size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
          <p>
            {parkActivity.stampCollectedAt
              ? `Stamp collected ${parkActivity.stampCollectedAt}`
              : 'Stamp not yet collected'}
          </p>
        </div>
      )}
      {
        park.bucketListItems &&
        park.bucketListItems.length > 0 && (
          <div className='top-0 flex'>
            {parkActivity.completedBucketListItems[0]?.id ? (
              <FaRegCheckSquare size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
            ) : (
              <FaRegSquare size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
            )}
            <p data-testid={'BLI'}>
              Bucket List Item:
              <br />
              {park.bucketListItems[0].task}
            </p>
          </div>
        )}
    </div>
  );
};

export default AchievementsView;
