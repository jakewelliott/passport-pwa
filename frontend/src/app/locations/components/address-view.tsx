import type { Address, Park } from '@/types';
import { useState } from 'react';
import { FiNavigation } from 'react-icons/fi';


const ParkAddress = ({ address }: { address: Address }) => {
	return (
		<div>
			<p>{address.title}</p>
			<p>{address.addressLineOne}</p>
			{address.addressLineTwo && <p>{address.addressLineTwo}</p>}
			<p>{address.city}, {address.state} {address.zipcode}</p>
		</div>
	);
};

export const AddressView = ({ park }: { park: Park }) => {

	const [showAll, setShowAll] = useState(false);
	const showButton = park.addresses && park.addresses.length > 1;

	const toggle = () => {
		setShowAll(!showAll);
	};

	const OtherAddresses = () =>
		park.addresses.slice(1).map((address) => (
			<ParkAddress key={address.title} address={address} />
		))


	const Button = () =>
		<button type='button' onClick={toggle} className='ml-4 inline-flex items-center'>
			{showAll ? 'Show Less' : 'Show More'}
		</button>

	return (
		<div className='top-0 flex' key={park.parkName}>
			<FiNavigation size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<ParkAddress address={park.addresses[0]} />
			{showAll && <OtherAddresses />}
			{showButton && <Button />}
		</div>
	);
};
