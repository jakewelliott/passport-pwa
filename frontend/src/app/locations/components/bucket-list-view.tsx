import { useBucketList } from '@/hooks/queries/useBucketList';
import { useStamp } from '@/hooks/queries/useStamps';
import DateHelper from '@/lib/date-helper';
import type { BucketListItem, CollectedStamp, Park } from '@/lib/mock/types';
import { FaRegCheckSquare, FaRegSquare, FaStamp } from 'react-icons/fa';

const BucketListItemView = ({ bucketListItem }: { bucketListItem: BucketListItem }) => {
	const isCompleted = bucketListItem.completedAt && !bucketListItem.deleted;
	const Checkbox = isCompleted ? FaRegCheckSquare : FaRegSquare;

	return (
		<div>
			<Checkbox size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<p data-testid={'BLI'}>
				Bucket List Item:
				<br />
				{bucketListItem.task}
			</p>
		</div>
	);
};

const StampView = ({ stamp }: { stamp: CollectedStamp | undefined }) => {
	return (
		<>
			<FaStamp size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<p>
				{stamp
					? `Stamp collected on ${DateHelper.stringify(new Date(stamp.createdAt)).replace(',', ' at')}`
					: 'Stamp not yet collected'}
			</p>
		</>
	);
};

export const BucketListView = ({ park }: { park: Park }) => {
	const { data: items, isLoading: itemsLoading } = useBucketList(park.id);
	const { data: stamp, isLoading: stampLoading } = useStamp(park.abbreviation);

	// TODO: replace with a placeholder
	if (itemsLoading || stampLoading) return null;

	return (
		<div className='flex flex-col gap-3' data-testid='achievements-view'>
			<div className='top-0 flex'>
				<StampView stamp={stamp} />
			</div>
			<div className='top-0 flex'>
				{items.map((item: BucketListItem) => (
					<BucketListItemView key={item.id} bucketListItem={item} />
				))}
			</div>
		</div>
	);
};
