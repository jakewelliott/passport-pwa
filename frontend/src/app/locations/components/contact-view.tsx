import type { Park } from '@/types';
import { FaRegEnvelope } from 'react-icons/fa';
import { FiNavigation, FiPhone } from 'react-icons/fi';

export const ContactView = ({ park }: { park: Park }) => {
    return (
        <div className='mx-4 flex flex-col gap-3' data-testid='location-contact'>
            {/* Coordinates */}
            <div className='top-0 flex'>
                <FiNavigation size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>
                    GPS: {park.coordinates.latitude.toFixed(4)}, {park.coordinates.longitude.toFixed(4)}
                </p>
            </div>
            {/* Phone */}
            <div className='top-0 flex'>
                <FiPhone size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>{String(park.phone).replace(/(\d{3})(\d{3})(\d{4})/, '($1) $2-$3')}</p>
            </div>
            {/* Email */}
            <div className='top-0 flex'>
                <FaRegEnvelope size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>{park.email}</p>
            </div>
        </div>
    );
};
