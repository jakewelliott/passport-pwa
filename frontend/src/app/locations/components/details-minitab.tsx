import type { Park } from '@/types';
import React from 'react';

export const DetailsMiniTab = ({ park }: { park: Park }) => {
	const renderTrails = (trails: string) => {
		return trails
			.split('\n')
			.filter((line) => line.trim() !== '') // Remove empty lines
			.map((line, index, array) => (
				<React.Fragment key={line}>
					{line}
					{index < array.length - 1 && <br />} {/* Add <br /> only if it's not the last line */}
				</React.Fragment>
			));
	};

	return (
		<div className='mt-6 mb-6 flex flex-col'>
			<p className='pr-6 pl-6'>
				{park.establishedYear && (
					<>
						<span className='green-text'>Established: </span>
						{park.establishedYear}
						<br />
						<br />
					</>
				)}
				{park.landmark && (
					<>
						<span className='green-text'>Landmark: </span>
						{park.landmark}
						<br />
						<br />
					</>
				)}
				{park.youCanFind && (
					<>
						<span className='green-text'>You can find... </span>
						{park.youCanFind}
						<br />
						<br />
					</>
				)}
				{park.trails && (
					<>
						<span className='green-text'>Trails: </span>
						<br />
						{renderTrails(park.trails)}
						<br />
						<br />
					</>
				)}
			</p>
			<div data-testid='icon-scroll-container' className='icon-scroll-container overflow-x-auto'>
				<div className='inline-flex gap-6 px-6'>
					{park.icons?.map((icon, index) => (
						<img
							src={`/icons/park/${icon.iconName}.svg`}
							width={55}
							height={55}
							key={icon.iconName}
							alt={`Park icon ${index + 1}`}
						/>
					))}
					<div className='w-px flex-shrink-0' />
				</div>
			</div>
		</div>
	);
};
