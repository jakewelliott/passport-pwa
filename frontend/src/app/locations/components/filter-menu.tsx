import { FaFilter } from 'react-icons/fa';

interface FilterMenuProps {
    searchQuery: string;
    setSearchQuery: (value: string) => void;
    setIsFilterModalOpen: (value: boolean) => void;
}

export const FilterMenu = ({ searchQuery, setSearchQuery, setIsFilterModalOpen }: FilterMenuProps) => (
    <div className='relative flex gap-2'>
        <div className='relative flex-1'>
            <input
                type='text'
                placeholder='Search parks...'
                value={searchQuery}
                onChange={(e) => setSearchQuery(e.target.value)}
                className='w-full rounded-lg border border-system-gray bg-supporting-lightblue p-3 pr-7 focus:border-secondary-darkteal focus:outline-none focus:ring-1 focus:ring-secondary-darkteal focus:ring-opacity-100'
            />
            {searchQuery && (
                <button
                    type='button'
                    onClick={() => setSearchQuery('')}
                    className='-translate-y-1/2 absolute top-1/2 right-3 transform bg-supporting-lightblue text-system-gray hover:text-secondary-darkteal'
                    aria-label='Clear search'
                >
                    &times;
                </button>
            )}
        </div>
        <button
            type='button'
            onClick={() => setIsFilterModalOpen(true)}
            className={
                'w-auto rounded-lg border border-system-gray bg-system-white px-3 py-2 transition-all duration-200'
            }
        >
            <FaFilter />
        </button>
    </div>
);
