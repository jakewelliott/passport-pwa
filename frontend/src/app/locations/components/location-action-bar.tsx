import type { Park } from '@/types';
import { FiGlobe, FiMail, FiNavigation, FiPhone } from 'react-icons/fi';

export const LocationActionBar = ({ park }: { park: Park }) => {
    // TODO: styling here needs to be fixed
    return (
        <div
            className='flex flex-row items-center justify-evenly bg-secondary-darkteal py-2.5'
            data-testid='location-action-bar'
        >
            {park.coordinates?.latitude && park.coordinates.longitude && (
                <a
                    href={`https://www.google.com/maps/place/${park.coordinates.latitude},${park.coordinates.longitude}`}
                    className='text-supporting-inactiveblue'
                >
                    <FiNavigation size={24} strokeWidth={3} />
                </a>
            )}
            {park.phone && (
                <a href={`tel://${park.phone}`} className='text-supporting-inactiveblue'>
                    <FiPhone size={24} strokeWidth={3} />
                </a>
            )}
            {park.website && (
                <a href={park.website} className='text-supporting-inactiveblue'>
                    <FiGlobe size={24} strokeWidth={3} />
                </a>
            )}
            {park.email && (
                <a href={`mailto:${park.email}`} className='text-supporting-inactiveblue'>
                    <FiMail size={24} strokeWidth={3} />
                </a>
            )}
        </div>
    );
};
