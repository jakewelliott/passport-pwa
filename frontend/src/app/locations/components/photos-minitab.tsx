import { ImageModal } from '@/components/image-modal';
import { a11yOnClick } from '@/lib/a11y';
import type { Park } from '@/types';
import { useState } from 'react';

const NoPhotos = () => {
	return <div className='p-4 text-center'>No photos available for this location.</div>;
};

export const PhotoGalleryMiniTab = ({ park }: { park: Park }) => {
	const [selectedPhoto, setSelectedPhoto] = useState<Park['photos'][0] | null>(null);

	if (!park.photos || park.photos.length === 0) return <NoPhotos />;

	const handleOpen = (photo: Park['photos'][0]) => {
		setSelectedPhoto(photo);
	};

	const handleClose = () => {
		setSelectedPhoto(null);
	};

	return (
		<>
			<div className='z-0 grid w-full grid-cols-3'>
				{park.photos.map((photo, index) => (
					<div key={photo.photoPath} className='aspect-square cursor-pointer' {...a11yOnClick(() => handleOpen(photo))}>
						<img
							data-testid={'photo'}
							src={`/photos/${photo.photoPath}`}
							alt={photo.alt || `Park photo ${index + 1}`}
							className='h-full w-full object-cover transition-opacity duration-300 hover:opacity-80'
						/>
					</div>
				))}
			</div>
			{selectedPhoto && <ImageModal photo={selectedPhoto} onClose={handleClose} />}
		</>
	);
};
