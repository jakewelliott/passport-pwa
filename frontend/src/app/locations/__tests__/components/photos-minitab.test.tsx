import { PhotoGalleryMiniTab } from '@/app/locations/components/photos-minitab';
import { mockPark } from '@/lib/testing/mock';
import { setupTestEnv } from '@/lib/testing/test-wrapper';
import { fireEvent, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it } from 'vitest';

const { render } = setupTestEnv();

const noPhotos = { ...mockPark, photos: [] };

describe('PhotoGalleryMiniTab', () => {
    beforeEach(() => {
        global.ResizeObserver = class ResizeObserver {
            observe() {}
            unobserve() {}
            disconnect() {}
        };
    });

    it('matches snapshot', () => {
        const { container } = render(<PhotoGalleryMiniTab park={mockPark} />);
        expect(container).toMatchSnapshot();
    });

    it('renders no photos message when photos array is empty', () => {
        render(<PhotoGalleryMiniTab park={noPhotos} />);
        expect(screen.getByText('No photos available for this location.')).toBeInTheDocument();
    });

    // ADAM: I chose not to delete these tests because they seem useful for edge cases

    it('renders all photos in a grid', () => {
        render(<PhotoGalleryMiniTab park={mockPark} />);
        const images = screen.getAllByTestId('photo');
        expect(images).toHaveLength(mockPark.photos.length);

        for (const [index, img] of images.entries()) {
            expect(img).toHaveAttribute('src', `/photos/${mockPark.photos[index].photoPath}`);
            expect(img).toHaveAttribute('alt', mockPark.photos[index].alt);
            expect(img).toHaveClass('h-full', 'w-full', 'object-cover');
        }
    });

    it('renders photos with fallback alt text when alt is not provided', () => {
        const photosWithoutAlt = [{ photoPath: '/test1.jpg', alt: '' }];
        const mockPark2 = { ...mockPark, photos: photosWithoutAlt };
        render(<PhotoGalleryMiniTab park={mockPark2} />);
        expect(screen.getByRole('img')).toHaveAttribute('alt', 'Park photo 1');
    });

    it('opens ImageModal when photo is clicked', () => {
        render(<PhotoGalleryMiniTab park={mockPark} />);
        const firstPhoto = screen.getAllByTestId('photo')[0];

        fireEvent.click(firstPhoto);

        // Check if modal is rendered
        const modalContainer = screen.getByRole('dialog', { hidden: true });
        expect(modalContainer).toBeInTheDocument();

        // Check if the correct image is displayed
        const modalImage = screen.getAllByRole('img').at(-1); // Get the last image (modal image)
        expect(modalImage).toHaveAttribute('src', `/photos/${mockPark.photos[0].photoPath}`);
        expect(modalImage).toHaveAttribute('alt', mockPark.photos[0].alt || 'Photo');
    });

    it('closes ImageModal when close button is clicked', () => {
        render(<PhotoGalleryMiniTab park={mockPark} />);

        // Open modal
        const firstPhoto = screen.getAllByRole('img')[0];
        fireEvent.click(firstPhoto);

        // Find and click close button (using the × symbol)
        const closeButton = screen.getByText('×');
        fireEvent.click(closeButton);

        // Verify modal is closed by checking that only the grid images remain
        const remainingImages = screen.getAllByRole('img');
        expect(remainingImages).toHaveLength(mockPark.photos.length);
    });
});
