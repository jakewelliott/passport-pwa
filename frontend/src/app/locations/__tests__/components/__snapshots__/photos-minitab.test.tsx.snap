// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PhotoGalleryMiniTab > matches snapshot 1`] = `
<div>
  <div
    class="z-0 grid w-full grid-cols-3"
  >
    <div
      class="aspect-square cursor-pointer"
    >
      <img
        alt="Chimney Rock State Park main photo"
        class="h-full w-full object-cover transition-opacity duration-300 hover:opacity-80"
        data-testid="photo"
        src="/photos/undefined"
      />
    </div>
  </div>
</div>
`;
